File: LeetCode/Easy/0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    ans = 0
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        def DFS(node):
            if not node:
                return 0
            l = DFS(node.left)
            r = DFS(node.right)
            self.ans = max(self.ans, l + r)
            return max(l, r) + 1
        DFS(root)
        return self.ans

File: LeetCode/Easy/0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int ret = 0;
    int diameterOfBinaryTree(TreeNode* root) {
        dfs(root);
        return ret;
    }
    
    int dfs(TreeNode* node){
        if (!node) return 0;
        
        int l = dfs(node->left);
        int r = dfs(node->right);
        ret = max(ret, l + r);
        return max(l, r) + 1;
    }
};